Add Sanic support

relud opened the pull request.

relud commented on src/dockerflow/sanic/app.py:
```diff
@@ -0,0 +1,306 @@
+
+    :param app: The Sanic app that this Dockerflow extension should be
+                initialized with.
+    :type root: ~sanic.Sanic or None
```
yep, definitely

relud commented on src/dockerflow/sanic/app.py:
```diff
@@ -0,0 +1,306 @@
+            self.init_app(app)
+        # Initialize the built-in checks.
+        if redis:
+            self.init_check(checks.check_redis_connected, redis)
```
I copied flask and ripped out what didn't work with minimal effort >_>

jezdez replied:
A+++ strategy 👍

relud commented on src/dockerflow/sanic/app.py:
```diff
@@ -0,0 +1,306 @@
+        extra = self.summary_extra(request)
+        extra['errno'] = 500
+        self.summary_logger.error(str(exception), extra=extra)
+        request['_logged'] = True
```
above in `_response_middleware` it prevents the info-level log of the same request without an exception message

jezdez replied:
Ah, d'oh, missed that. Thanks :)

relud commented on src/dockerflow/sanic/app.py:
```diff
@@ -0,0 +1,306 @@
+            'lang': request.headers.get('Accept-Language', ''),
+            'method': request.method,
+            'path': request.path,
+            'uid': '',
```
`''` was the default if there was no user logged in via `flask_login`. there's `sanic_auth` but it was taking up too much time to figure out a reasonable way to detect if it was active on the app so i removed support for now.

jezdez replied:
Gotcha, we can pick this up in case auth is needed in a Sanic project then. Would you mind creating a ticket?

relud replied:
added support for pulling `uid` using `sanic_auth`

relud replied:
er, actually i removed pulling `uid` from `sanic_auth` because it doesn't actually provide auth, it just provides a framework for adding your own auth, and that made it non-trivial to test. I'll open a ticket.

jezdez commented:
> > One thing I wondered though is the code overlaps with the Flask support, do you think we could/should abstract those away?
> 
> I don't think we should abstract them away. My experience switching from flask to sanic was that it's similar, but definitely not trying to be api compatible at all (many things slightly renamed). I think that having sanic and flask separate allows for properly supporting each in ways that make the most sense for them (such as adding await and async keywords, which I haven't figured out how to use yet, but I know is important)

Cool, that makes sense to me. Thanks for elaborating!

relud commented:
@jezdez okay, this should be ready for a final review now

jezdez commented on src/dockerflow/sanic/checks.py:
```diff
@@ -0,0 +1,68 @@
+    errors = []
+
+    try:
+        result = await app.redis.ping()
```
Expecting `app.redis` is a bad assumption and we should follow the idea of simply coupling the check to one specific Sanic extension, such as [sanic-redis](https://github.com/strahe/sanic-redis). See above for more details.

relud commented on src/dockerflow/flask/checks.py:
```diff
@@ -4,11 +4,7 @@
-    DEBUG, INFO, WARNING, ERROR, CRITICAL, STATUSES, level_to_text,
-    CheckMessage, Debug, Info, Warning, Error, Critical,
-)
+from .. import health, messages
```
ah, good point. i moved everything back into place there, and imported all that same stuff into `dockerflow/sanic/checks.py` to provide a similar interface.

relud commented on src/dockerflow/sanic/app.py:
```diff
@@ -0,0 +1,324 @@
+
+    :param aioredis: A boolean indicating that app.redis should be used by the
+                     built-in Dockerflow check for the aioredis connection.
+    :type aioredis: bool
```
okay, i made it a sanic redis instance. also, i found a bug in the process that was making it not work at all, which should be fixed now.

relud commented:
> Sorry but this needs another pass

no worries, my bad for misspeaking. `s/final/another/`

ready again.

relud commented:
ready for another review.

I was messing with tests and I noticed that `from dockerflow.flask.checks.messages import *` created a dependency on `flask`, so I moved `dockerflow.flask.checks.messages` to `dockerflow.checks.messages` and then imported all of the values into the old location and threw a `PendingDeprecationWarning`, so `dockerflow.sanic` no longer depends on `flask`. While I was at it I added an `extras_require` for `sanic`.

relud commented:
bump @jezdez
